File: contrib/corpus_hash.cpp
// The independent C++ arm of the corpus differential: reads image files with
// nothing but public OpenImageIO C++ API — none of oiio-sys' shims — and
// prints one line per subimage: path, subimage index, dimensions, and an
// FNV-1a 64 hash of the pixels converted to float. `examples/corpus_hash.rs`
// is the crate arm printing the identical format; diff the outputs.
//
// Reads file paths from stdin, one per line. Deep subimages print DEEP (the
// deep API is compared elsewhere); files or subimages that fail print ERROR.
//
// Build (MSVC, against the same vcpkg OpenImageIO the crate links):
//   cl /std:c++17 /EHsc /utf-8 /MD corpus_hash.cpp
//      /I C:\vcpkg\installed\x64-windows\include
//      /link /LIBPATH:C:\vcpkg\installed\x64-windows\lib
//      OpenImageIO.lib OpenImageIO_Util.lib
// Run with C:\vcpkg\installed\x64-windows\bin on PATH.

#include <cstdint>
#include <cstdio>
#include <iostream>
#include <string>
#include <vector>

#include <OpenImageIO/imageio.h>

static uint64_t
fnv1a64(const unsigned char* data, size_t len)
{
    uint64_t hash = 14695981039346656037ull;
    for (size_t i = 0; i < len; ++i) {
        hash ^= data[i];
        hash *= 1099511628211ull;
    }
    return hash;
}

int
main()
{
    std::string path;
    while (std::getline(std::cin, path)) {
        if (path.empty())
            continue;
        auto input = OIIO::ImageInput::open(path);
        if (!input) {
            // Drain the message so OpenImageIO's at-exit "pending error"
            // dump cannot land in the output being diffed.
            (void)OIIO::geterror(true);
            std::printf("%s\t-\tERROR\n", path.c_str());
            continue;
        }
        for (int s = 0;; ++s) {
            if (!input->seek_subimage(s, 0))
                break;
            const OIIO::ImageSpec& spec = input->spec();
            if (spec.deep) {
                std::printf("%s\t%d\tDEEP\n", path.c_str(), s);
                continue;
            }
            const uint64_t values = spec.image_pixels()
                                    * uint64_t(spec.nchannels);
            // The corpora's largest flat files are far below this; anything
            // over it would only slow the sweep down.
            if (values > (uint64_t(1) << 28)) {
                std::printf("%s\t%d\tSKIPPED-LARGE\n", path.c_str(), s);
                continue;
            }
            std::vector<float> pixels(static_cast<size_t>(values));
            if (!input->read_image(s, 0, 0, -1, OIIO::TypeFloat,
                                   pixels.data())) {
                (void)input->geterror(true);
                std::printf("%s\t%d\tERROR\n", path.c_str(), s);
                continue;
            }
            const unsigned char* bytes
                = reinterpret_cast<const unsigned char*>(pixels.data());
            const size_t nbytes = pixels.size() * sizeof(float);
            const uint64_t hash = fnv1a64(bytes, nbytes);
            std::printf("%s\t%d\t%dx%dx%d\t%016llx\n", path.c_str(), s,
                        spec.width, spec.height, spec.nchannels,
                        (unsigned long long)hash);
        }
        input->close();
    }
    return 0;
}

File: examples/corpus_hash.rs
//! The crate arm of the corpus differential: reads image files through the
//! safe `oiio` API and prints one line per subimage — path, subimage index,
//! dimensions, and an FNV-1a 64 hash of the pixels as `f32` — in exactly the
//! format of `contrib/corpus_hash.cpp`, the standalone C++ arm that uses no
//! part of this crate. Feed both the same file list on stdin and diff.

use oiio::{Error, ImageInput};
use std::io::BufRead;
use std::path::Path;

fn fnv1a64(data: &[u8]) -> u64 {
    let mut hash = 0xcbf2_9ce4_8422_2325_u64;
    for &byte in data {
        hash ^= u64::from(byte);
        hash = hash.wrapping_mul(0x0000_0100_0000_01b3);
    }
    hash
}

fn main() {
    let stdin = std::io::stdin();
    for line in stdin.lock().lines() {
        let Ok(path) = line else { break };
        if path.is_empty() {
            continue;
        }
        let Ok(mut input) = ImageInput::from_path(Path::new(&path)) else {
            println!("{path}\t-\tERROR");
            continue;
        };
        for subimage in 0_u32.. {
            let spec = match input.image_spec_at(subimage, 0) {
                Ok(spec) => spec,
                Err(Error::InvalidImageLevel { .. }) => break,
                Err(_) => {
                    println!("{path}\t{subimage}\tERROR");
                    break;
                }
            };
            if spec.is_deep() {
                println!("{path}\t{subimage}\tDEEP");
                continue;
            }
            let [width, height, _] = spec.dimensions();
            let channels = spec.channel_count();
            let Ok(values) = spec.element_count() else {
                println!("{path}\t{subimage}\tERROR");
                continue;
            };
            // Mirror of the C++ arm's large-file skip.
            if values as u64 > 1 << 28 {
                println!("{path}\t{subimage}\tSKIPPED-LARGE");
                continue;
            }
            let mut pixels = vec![0.0_f32; values];
            match input.read_image_into_at(subimage, 0, &mut pixels) {
                Ok(()) => {
                    let bytes: &[u8] = bytemuck_bytes(&pixels);
                    let hash = fnv1a64(bytes);
                    println!("{path}\t{subimage}\t{width}x{height}x{channels}\t{hash:016x}");
                }
                Err(_) => println!("{path}\t{subimage}\tERROR"),
            }
        }
        let _ = input.close();
    }
}

/// The float buffer viewed as bytes, dependency-free.
fn bytemuck_bytes(pixels: &[f32]) -> &[u8] {
    // SAFETY: f32 has no padding and any byte pattern is readable; the slice
    // covers exactly the same memory.
    unsafe {
        std::slice::from_raw_parts(pixels.as_ptr().cast::<u8>(), std::mem::size_of_val(pixels))
    }
}
